File: mobile/lib/models/stock.dart
import 'package:equatable/equatable.dart';
import 'product.dart';
import 'location.dart';
import 'storage_box.dart';

class Stock extends Equatable {
  final int id;
  final int productId;
  final Product? product;
  final int locationId;
  final Location? location;
  final int storageBoxId;
  final StorageBox? storageBox;
  final String serialNumber;
  final String status; // available, reserved, sold, transfer
  final double buyPrice;
  final double sellPrice;
  final String notes;
  final String? supplierName;
  final DateTime? receivedAt;
  final DateTime? soldAt;
  final int? transactionId;
  final bool barcodePrinted;
  final DateTime? barcodePrintedAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Stock({
    required this.id,
    required this.productId,
    this.product,
    required this.locationId,
    this.location,
    required this.storageBoxId,
    this.storageBox,
    required this.serialNumber,
    this.status = 'available',
    required this.buyPrice,
    required this.sellPrice,
    this.notes = '',
    this.supplierName,
    this.receivedAt,
    this.soldAt,
    this.transactionId,
    this.barcodePrinted = false,
    this.barcodePrintedAt,
    this.createdAt,
    this.updatedAt,
  });

  factory Stock.fromJson(Map<String, dynamic> json) {
    return Stock(
      id: json['id'] ?? 0,
      productId: json['product_id'] ?? 0,
      product: json['product'] != null ? Product.fromJson(json['product']) : null,
      locationId: json['location_id'] ?? 0,
      location: json['location'] != null ? Location.fromJson(json['location']) : null,
      storageBoxId: json['storage_box_id'] ?? 0,
      storageBox: json['storage_box'] != null ? StorageBox.fromJson(json['storage_box']) : null,
      serialNumber: json['serial_number'] ?? '',
      status: json['status'] ?? 'available',
      buyPrice: (json['buy_price'] ?? 0).toDouble(),
      sellPrice: (json['sell_price'] ?? json['selling_price'] ?? 0).toDouble(),
      notes: json['notes'] ?? '',
      supplierName: json['supplier_name'],
      receivedAt: json['received_at'] != null ? DateTime.parse(json['received_at']) : null,
      soldAt: json['sold_at'] != null ? DateTime.parse(json['sold_at']) : null,
      transactionId: json['transaction_id'],
      barcodePrinted: json['barcode_printed'] ?? false,
      barcodePrintedAt: json['barcode_printed_at'] != null
          ? DateTime.parse(json['barcode_printed_at'])
          : null,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  // Alias for sellPrice
  double get sellingPrice => sellPrice;
  double get weight => product?.weight ?? 0;

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'product_id': productId,
      'location_id': locationId,
      'storage_box_id': storageBoxId,
      'serial_number': serialNumber,
      'status': status,
      'buy_price': buyPrice,
      'sell_price': sellPrice,
      'notes': notes,
      'supplier_name': supplierName,
      'barcode_printed': barcodePrinted,
    };
  }

  Stock copyWith({
    int? id,
    int? productId,
    Product? product,
    int? locationId,
    Location? location,
    int? storageBoxId,
    StorageBox? storageBox,
    String? serialNumber,
    String? status,
    double? buyPrice,
    double? sellPrice,
    String? notes,
    String? supplierName,
    DateTime? receivedAt,
    DateTime? soldAt,
    int? transactionId,
    bool? barcodePrinted,
    DateTime? barcodePrintedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Stock(
      id: id ?? this.id,
      productId: productId ?? this.productId,
      product: product ?? this.product,
      locationId: locationId ?? this.locationId,
      location: location ?? this.location,
      storageBoxId: storageBoxId ?? this.storageBoxId,
      storageBox: storageBox ?? this.storageBox,
      serialNumber: serialNumber ?? this.serialNumber,
      status: status ?? this.status,
      buyPrice: buyPrice ?? this.buyPrice,
      sellPrice: sellPrice ?? this.sellPrice,
      notes: notes ?? this.notes,
      supplierName: supplierName ?? this.supplierName,
      receivedAt: receivedAt ?? this.receivedAt,
      soldAt: soldAt ?? this.soldAt,
      transactionId: transactionId ?? this.transactionId,
      barcodePrinted: barcodePrinted ?? this.barcodePrinted,
      barcodePrintedAt: barcodePrintedAt ?? this.barcodePrintedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [id, productId, serialNumber, status, locationId];
}

File: backend/models/stock.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// StockStatus defines the status of stock item
type StockStatus string

const (
	StockStatusAvailable StockStatus = "available" // Ready for sale
	StockStatusReserved  StockStatus = "reserved"  // Reserved for order
	StockStatusSold      StockStatus = "sold"      // Already sold
	StockStatusTransfer  StockStatus = "transfer"  // In transfer between locations
)

// Stock represents individual stock item with location tracking
// Harga tidak disimpan di sini - selalu dihitung dari gold_category.sell_price * product.weight
type Stock struct {
	ID           uint           `gorm:"primarykey" json:"id"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`
	ProductID    uint           `gorm:"not null;index" json:"product_id"`
	Product      Product        `gorm:"foreignKey:ProductID" json:"product,omitempty"`
	LocationID   uint           `gorm:"not null;index" json:"location_id"`
	Location     Location       `gorm:"foreignKey:LocationID" json:"location,omitempty"`
	StorageBoxID uint           `gorm:"not null;index" json:"storage_box_id"`
	StorageBox   StorageBox     `gorm:"foreignKey:StorageBoxID" json:"storage_box,omitempty"`
	SerialNumber string         `gorm:"not null;size:50" json:"serial_number"` // Unique identifier for each piece - unique index created manually in migration
	Status       StockStatus    `gorm:"not null;size:20;default:'available';index" json:"status"`
	Notes        string         `gorm:"size:255" json:"notes"`

	// Source tracking
	SupplierName string     `gorm:"size:100" json:"supplier_name,omitempty"` // Name of supplier/distributor
	ReceivedAt   *time.Time `json:"received_at,omitempty"`                   // When received at location

	// Sales tracking
	SoldAt        *time.Time `json:"sold_at,omitempty"`
	TransactionID *uint      `gorm:"index" json:"transaction_id,omitempty"`

	// Barcode printing tracking
	BarcodePrinted   bool       `gorm:"default:false" json:"barcode_printed"`
	BarcodePrintedAt *time.Time `json:"barcode_printed_at,omitempty"`
}

// StockTransfer represents stock movement between locations
type StockTransfer struct {
	ID              uint           `gorm:"primarykey" json:"id"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
	TransferNumber  string         `gorm:"not null;size:50" json:"transfer_number"` // unique index created manually in migration
	StockID         uint           `gorm:"not null;index" json:"stock_id"`
	Stock           Stock          `gorm:"foreignKey:StockID" json:"stock,omitempty"`
	FromLocationID  uint           `gorm:"not null;index" json:"from_location_id"`
	FromLocation    Location       `gorm:"foreignKey:FromLocationID" json:"from_location,omitempty"`
	FromBoxID       uint           `gorm:"not null;index" json:"from_box_id"`
	FromBox         StorageBox     `gorm:"foreignKey:FromBoxID" json:"from_box,omitempty"`
	ToLocationID    uint           `gorm:"not null;index" json:"to_location_id"`
	ToLocation      Location       `gorm:"foreignKey:ToLocationID" json:"to_location,omitempty"`
	ToBoxID         uint           `gorm:"not null;index" json:"to_box_id"`
	ToBox           StorageBox     `gorm:"foreignKey:ToBoxID" json:"to_box,omitempty"`
	TransferredByID uint           `gorm:"not null;index" json:"transferred_by_id"`
	TransferredBy   User           `gorm:"foreignKey:TransferredByID" json:"transferred_by,omitempty"`
	TransferredAt   time.Time      `json:"transferred_at"`
	Notes           string         `gorm:"size:255" json:"notes"`
	Status          string         `gorm:"not null;size:20;default:'pending'" json:"status"` // pending, completed, cancelled
}
